File: src/responses/components.rs
//! Message components used in the registration process and command replies.

use poise::serenity_prelude as serenity;

use serenity::*;

/// The ID of the `Continue` button for starting registration.
///
/// See [`instructions_continue_button`].
pub const INSTRUCTIONS_CONTINUE: &str = "instructions-continue";

/// A `Continue` button for starting registration.
///
/// See [`INSTRUCTIONS_CONTINUE`].
pub fn instructions_continue_button() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents
{
    button_continue(INSTRUCTIONS_CONTINUE)
}

/// A link button that opens direct messages.
pub fn instructions_sent_button() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    |comp| {
        comp.create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("Open direct messages")
                    .url("https://discordapp.com/channels/@me")
                    .style(ButtonStyle::Link)
            })
        })
    }
}

/// The ID of the `Yes` button when confirming a name.
///
/// See [`confirm_name_buttons`].
pub const NAME_CONFIRM_YES: &str = "name-confirm-yes";

/// The ID of the `No` button when confirming a name.
///
/// See [`confirm_name_buttons`].
pub const NAME_CONFIRM_NO: &str = "name-confirm-no";

/// `Yes` and `No` buttons for confirming a [`User`]'s name.
///
/// See [`NAME_CONFIRM_YES`] and [`NAME_CONFIRM_NO`].
pub fn confirm_name_buttons() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    buttons_yes_no("Yes", "No", NAME_CONFIRM_YES, NAME_CONFIRM_NO)
}

/// The ID of the `Yes` button when confirming a [`VerifiedUserKind`].
///
/// See [`confirm_kind_buttons`].
///
/// [`VerifiedUserKind`]: crate::prelude::VerifiedUserKind
pub const KIND_CONFIRM_YES: &str = "kind-confirm-yes";

/// The ID of the `No` button when confirming a [`VerifiedUserKind`].
///
/// [`VerifiedUserKind`]: crate::prelude::VerifiedUserKind
///
/// See [`confirm_kind_buttons`].
pub const KIND_CONFIRM_NO: &str = "kind-confirm-no";

/// `Yes` and `No` buttons for confirming a [`VerifiedUserKind`].
///
/// See [`KIND_CONFIRM_YES`] and [`KIND_CONFIRM_NO`].
///
/// [`VerifiedUserKind`]: crate::prelude::VerifiedUserKind
pub fn confirm_kind_buttons() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    buttons_yes_no("Yes", "No", KIND_CONFIRM_YES, KIND_CONFIRM_NO)
}

/// The ID of the `Continue` button after verification.
///
/// See [`verified_continue_button`].
pub const VERIFIED_CONTINUE: &str = "verified-continue";

/// A `Continue` button for proceeding past verification to the optional questions.
///
/// See [`VERIFIED_CONTINUE`].
pub fn verified_continue_button() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    button_continue(VERIFIED_CONTINUE)
}

/// The ID of the `He / Him` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_HE_HIM: &str = "pronouns-he-him";

/// The ID of the `She / Her` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_SHE_HER: &str = "pronouns-she-her";

/// The ID of the `They / Them` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_THEY_THEM: &str = "pronouns-they-them";

/// The ID of the `Xe / Xem` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_XE_XEM: &str = "pronouns-xe-xem";

/// The ID of the `Any` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_ANY: &str = "pronouns-any";

/// The ID of the `Ask me` pronouns button.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_ASK: &str = "pronouns-ask";

/// The ID of the `Skip` button for pronouns.
///
/// See [`pronouns_buttons`].
pub const PRONOUNS_SKIP: &str = "pronouns-skip";

/// Buttons for selecting (or skipping) pronouns.
///
/// See [`PRONOUNS_HE_HIM`], [`PRONOUNS_SHE_HER`], [`PRONOUNS_THEY_THEM`], [`PRONOUNS_XE_XEM`], [`PRONOUNS_ANY`], [`PRONOUNS_ASK`], and [`PRONOUNS_SKIP`].
pub fn pronouns_buttons() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    |comp| {
        comp.create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("He / Him")
                    .custom_id(PRONOUNS_HE_HIM)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("She / Her")
                    .custom_id(PRONOUNS_SHE_HER)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("They / Them")
                    .custom_id(PRONOUNS_THEY_THEM)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Xe / Xem")
                    .custom_id(PRONOUNS_XE_XEM)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Any")
                    .custom_id(PRONOUNS_ANY)
                    .style(ButtonStyle::Primary)
            })
        })
        // NOTE: An action row can only hold up to five buttons
        .create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("Ask me")
                    .custom_id(PRONOUNS_ASK)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Skip")
                    .custom_id(PRONOUNS_SKIP)
                    .style(ButtonStyle::Danger)
            })
        })
    }
}

/// The ID of the `Catered halls (Jubilee)` housing button.
///
/// See [`housing_buttons`].
pub const HOUSING_JC_CATERED: &str = "housing-jc-catered";

/// The ID of the `Self-atered halls (Jubilee area)` housing button.
///
/// See [`housing_buttons`].
pub const HOUSING_JC_SELF_CATERED: &str = "housing-jc-self-catered";

/// The ID of the `Catered halls (University Park)` housing button.
///
/// See [`housing_buttons`].
pub const HOUSING_UP_CATERED: &str = "housing-up-catered";

/// The ID of the `Self-atered halls (University Park area)` housing button.
///
/// See [`housing_buttons`].
pub const HOUSING_UP_SELF_CATERED: &str = "housing-up-self-catered";

/// The ID of the `Private housing` housing button.
///
/// See [`housing_buttons`].
pub const HOUSING_PRIVATE: &str = "housing-private";

/// The ID of the `Skip` button for housing.
///
/// See [`housing_buttons`].
pub const HOUSING_SKIP: &str = "housing-skip";

/// Buttons for selecting (or skipping) housing.
///
/// See [`HOUSING_JC_CATERED`], [`HOUSING_JC_SELF_CATERED`], [`HOUSING_UP_CATERED`], [`HOUSING_UP_SELF_CATERED`], [`HOUSING_PRIVATE`], and [`HOUSING_SKIP`].
pub fn housing_buttons() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    |comp| {
        comp.create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("Catered halls (Jubilee)")
                    .custom_id(HOUSING_JC_CATERED)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Catered halls (University Park)")
                    .custom_id(HOUSING_UP_CATERED)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Self-catered halls (Jubilee area)")
                    .custom_id(HOUSING_JC_SELF_CATERED)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Self-catered halls (University Park area)")
                    .custom_id(HOUSING_UP_SELF_CATERED)
                    .style(ButtonStyle::Primary)
            })
            .create_button(|button| {
                button
                    .label("Private housing")
                    .custom_id(HOUSING_PRIVATE)
                    .style(ButtonStyle::Primary)
            })
        })
        .create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("Skip")
                    .custom_id(HOUSING_SKIP)
                    .style(ButtonStyle::Danger)
            })
        })
    }
}

/// The ID of the `Approve` nickname button.
///
/// See [`nickname_approval_buttons`].
pub const NICKNAME_APPROVE: &str = "nickname-approve";

/// The ID of the `Deny` nickname button.
///
/// See [`nickname_approval_buttons`].
pub const NICKNAME_DENY: &str = "nickname-deny";

/// `Approve` and `Deny` buttons for nickname approval.
///
/// See [`NICKNAME_APPROVE`] and [`NICKNAME_DENY`].
pub fn nickname_approval_buttons() -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    buttons_yes_no("Approve", "Deny", NICKNAME_APPROVE, NICKNAME_DENY)
}

fn button_continue(
    id: &'static str,
) -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    move |comp| {
        comp.create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label("Continue")
                    .custom_id(id)
                    .style(ButtonStyle::Primary)
            })
        })
    }
}

fn buttons_yes_no(
    yes_text: &'static str,
    no_text: &'static str,
    yes_id: &'static str,
    no_id: &'static str,
) -> impl FnOnce(&mut CreateComponents) -> &mut CreateComponents {
    move |comp| {
        comp.create_action_row(|row| {
            row.create_button(|button| {
                button
                    .label(yes_text)
                    .custom_id(yes_id)
                    .style(ButtonStyle::Success)
            })
            .create_button(|button| {
                button
                    .label(no_text)
                    .custom_id(no_id)
                    .style(ButtonStyle::Danger)
            })
        })
    }
}

File: norris/responses/components.py
"""
Message components used in the registration process and command replies.
"""

from discord import ButtonStyle, Interaction
from discord.ui import Button, View, button

from ..bot import Norris
from ..config import Housing, Pronouns
from ..events.component_interaction import (
    housing,
    instructions,
    kind_confirm,
    name_confirm,
    pronouns,
    verified,
    nickname,
)


class InstructionsContinueView(View):
    """
    A view containing a `Continue` button for starting registration.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Continue", style=ButtonStyle.primary)
    async def continue_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `Continue` button handler for starting registration.
        """
        await instructions.continue_clicked(interaction, self._norris)


class OpenDirectMessagesView(View):
    """
    A view containing a link button that opens direct messages.
    """

    def __init__(self) -> None:
        super().__init__()

        # Add link button to open DMs
        self.add_item(open_dm_button())


def open_dm_button() -> Button:
    """
    A link button that opens direct messages.
    """
    return Button(label="Open direct messages",
                  style=ButtonStyle.link,
                  url="https://discord.com/channels/@me")


class NameConfirmView(View):
    """
    A view containing `Yes` and `No` buttons for confirming a user's name.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Yes", style=ButtonStyle.green)
    async def yes_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `Yes` button handler to confirm a user's name.
        """
        await name_confirm.yes_clicked(interaction, self._norris)

    @button(label="No", style=ButtonStyle.red)
    async def no_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `No` button handler to un-confirm a user's name.
        """
        await name_confirm.no_clicked(interaction, self._norris)


class KindConfirmView(View):
    """
    A view containing `Yes` and `No` buttons to confirm a user's
    `norris.model.verified_user.VerifiedUserKind`.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Yes", style=ButtonStyle.green)
    async def yes_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `Yes` button handler to confirm a user's kind.
        """
        await kind_confirm.yes_clicked(interaction, self._norris)

    @button(label="No", style=ButtonStyle.red)
    async def no_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `No` button handler to indicate an incorrectly detected kind.
        """
        await kind_confirm.no_clicked(interaction, self._norris)


class VerifiedContinueView(View):
    """
    A view containing a `Continue` button for proceeding past verification to the
    optional questions.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Continue", style=ButtonStyle.primary)
    async def continue_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        A `Continue` button handler for proceeding past verification to the optional
        questions.
        """
        await verified.continue_clicked(interaction, self._norris)


class PronounsView(View):
    """
    A view containing buttons for selecting (or skipping) pronouns.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="He / Him", style=ButtonStyle.primary)
    async def he_him_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `He / Him` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.HE_HIM, self._norris)

    @button(label="She / Her", style=ButtonStyle.primary)
    async def she_her_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `She / Her` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.SHE_HER, self._norris)

    @button(label="They / Them", style=ButtonStyle.primary)
    async def they_them_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `They / Them` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.THEY_THEM, self._norris)

    @button(label="Xe / Xem", style=ButtonStyle.primary)
    async def xe_xem_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Xe / Xem` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.XE_XEM, self._norris)

    @button(label="Any", style=ButtonStyle.primary)
    async def any_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Any` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.ANY, self._norris)

    @button(label="Ask me", style=ButtonStyle.primary)
    async def ask_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Ask me` pronouns button handler.
        """
        await pronouns.pronouns_clicked(interaction, Pronouns.ASK, self._norris)

    @button(label="Skip", style=ButtonStyle.red)
    async def skip_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Skip` pronouns button handler.
        """
        await pronouns.skip_clicked(interaction, self._norris)


class HousingView(View):
    """
    A view containing buttons for selecting (or skipping) housing.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Catered halls (Jubilee)", style=ButtonStyle.primary)
    async def jc_catered_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Catered halls (Jubilee)` housing button handler.
        """
        await housing.housing_clicked(interaction, Housing.JC_CATERED, self._norris)

    @button(label="Catered halls (University Park)", style=ButtonStyle.primary)
    async def up_catered_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Catered halls (University Park)` housing button handler.
        """
        await housing.housing_clicked(interaction, Housing.UP_CATERED, self._norris)

    @button(label="Self-catered halls (Jubilee area)", style=ButtonStyle.primary)
    async def jc_self_catered_clicked(self,
                                      _: Button,
                                      interaction: Interaction) -> None:
        """
        The `Self-catered halls (Jubilee)` housing button handler.
        """
        await housing.housing_clicked(interaction,
                                      Housing.JC_SELF_CATERED,
                                      self._norris)

    @button(label="Self-catered halls (University Park area)",
            style=ButtonStyle.primary)
    async def up_self_catered_clicked(self,
                                      _: Button,
                                      interaction: Interaction) -> None:
        """
        The `Self-catered halls (University Park)` housing button handler.
        """
        await housing.housing_clicked(interaction,
                                      Housing.UP_SELF_CATERED,
                                      self._norris)

    @button(label="Private housing", style=ButtonStyle.primary)
    async def private_house_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Private house` housing button handler.
        """
        await housing.housing_clicked(interaction, Housing.PRIVATE_HOUSE, self._norris)

    @button(label="Skip", style=ButtonStyle.red)
    async def skip_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Skip` housing button handler.
        """
        await housing.skip_clicked(interaction, self._norris)


class NicknameView(View):
    """
    A view containing buttons to approve or deny nicknames.
    """

    _norris: Norris

    def __init__(self, norris: Norris) -> None:
        super().__init__()
        self._norris = norris

    @button(label="Approve", style=ButtonStyle.green)
    async def approve_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Approve` nickname button handler.
        """
        user_id = 0  # TODO: figure out which user requested it and their nickname
        new_nickname = ""
        await nickname.approve_clicked(user_id, new_nickname, interaction, self._norris)

    @button(label="Deny", style=ButtonStyle.red)
    async def deny_clicked(self, _: Button, interaction: Interaction) -> None:
        """
        The `Deny` nickname button handler.
        """
        new_nickname = ""  # TODO: figure out how to get the requested nickname
        await nickname.deny_clicked(new_nickname, interaction)
